src/Database/Schema/SqlGeneratorInterface.php

```diff
@@ -29,44 +29,44 @@ interface SqlGeneratorInterface
      * Uses the connection to access the schema dialect
      * to generate platform specific SQL.
      *
-     * @param \Cake\Database\Connection $connection The connection to generate SQL for.
+     * @param \Cake\Datasource\ConnectionInterface $connection The connection to generate SQL for.
      * @return array List of SQL statements to create the table and the
      *    required indexes.
      */
-    public function createSql(Connection $connection): array;
+    public function createSql(ConnectionInterface $connection): array;
```

Comment:
Why is this necessary? `ConnectionInterface` is too generic here. The `Database\Connection` class is relational db centric and provides the methods and drivers which use SQL. Since this interface is also specific to generating SQL for relational db, typhinting with `Connection` is the right thing to do.

In general all classes under `Database` package should just use `Connection` for typehint as they all relate to relational database and  `ConnectionInterface` doesn't have all the requisite methods.

We could add a new `Database\ConnectionInterface` which extends `Datasource\ConnectionInterface` but it wouldn't serve any purpose as no one is going to write an alternate implementation for `Database\Connection`.

Reply:
I would in fact remove the `disableConstraints()` method from `ConnectionInterface` as it would only be applicable for relational databases.

Reply:
As seen in the connection class of elastic search plugin, the method had to be made into a no-op 

https://github.com/cakephp/elastic-search/blob/c1394953fd09f38d157092f86760fac653c31b41/src/Datasource/Connection.php#L167-L173

Reply:
`ConnectionInterface::setLogger()` and `ConnectionInterface::getLogger()` use `\Cake\Database\Log\QueryLogger` for typehinting which too is technically incorrect since the `database` package is not a dependency of `datasource` packages. It's the other way round.

So `Datasource\ConnectionInterface` serves no real purpose. We would be actually lying by using it for typehints. We should directly use the `Database\Connection`. I would suggest simply dropping `ConnectionInterface`.

The `ConnectionManager` and `ConnectionRegistry` should be able to take any object as connection instance, which they already do. For e.g. it these classes load instances of `Muffin\Webservice\Connection` which doesn't even implement `ConnectionInterface` and that behavior should be preserved.